From 681185fbe4d402ae41b63d222abd8384f06e2b56 Mon Sep 17 00:00:00 2001
From: Hao Bui <hao.bui.yg@renesas.com>
Date: Wed, 26 Apr 2023 10:03:51 +0700
Subject: [PATCH 2/3] rzg3s: Add SMC command for PCIe RST_RSM_B terminal
 control

Add SMC command to support PCIe RST_RSM_B control. Example usage from
Linux kernel as following:

a.Set RST_RSM_B=1 after PCIe power is applied
 arm_smccc_smc(RZ_SIP_SVC_SET_PCIE_RST_RSMB, 0xd74, 0x1, 0, 0, 0, 0, 0, &local_res);
b.Set RST_RSM_B=1 before turning off the power supply
 arm_smccc_smc(RZ_SIP_SVC_SET_PCIE_RST_RSMB, 0xd74, 0x0, 0, 0, 0, 0, 0, &local_res);

Signed-off-by: Hao Bui <hao.bui.yg@renesas.com>
---
 plat/renesas/rz/common/include/rz_sip_svc.h  |  3 +++
 plat/renesas/rz/common/rz_plat_sip_handler.c | 13 +++++++++++++
 plat/renesas/rz/soc/g3s/include/rz_soc_def.h |  1 +
 3 files changed, 17 insertions(+)

diff --git a/plat/renesas/rz/common/include/rz_sip_svc.h b/plat/renesas/rz/common/include/rz_sip_svc.h
index 62ed2b3fd..542e33a1f 100644
--- a/plat/renesas/rz/common/include/rz_sip_svc.h
+++ b/plat/renesas/rz/common/include/rz_sip_svc.h
@@ -13,4 +13,7 @@
 /* Function ID to get Chip ID */
 #define RZ_SIP_SVC_GET_CHIPID		U(0x82000011)
 
+/* Function ID to set PCIE RST_RSM_B */
+#define RZ_SIP_SVC_SET_PCIE_RST_RSMB		U(0x82000013)
+
 #endif /* __RZ_SIP_SVC_H__ */
diff --git a/plat/renesas/rz/common/rz_plat_sip_handler.c b/plat/renesas/rz/common/rz_plat_sip_handler.c
index 1c4b612c1..9d2e976b3 100644
--- a/plat/renesas/rz/common/rz_plat_sip_handler.c
+++ b/plat/renesas/rz/common/rz_plat_sip_handler.c
@@ -48,6 +48,17 @@ static uintptr_t rz_otp_handler_chipid(void *handle, u_register_t x1, u_register
 	SMC_RET4(handle, chipid[0], chipid[1], chipid[2], chipid[3]);
 }
 
+static uintptr_t rz_otp_handler_set_pcie(void *handle, u_register_t x1, u_register_t x2)
+{
+	if ((uintptr_t)x1 == RZG3S_SYSC_PCIE_RST_RSM_B_OFFSET) {
+		mmio_write_32(RZG3S_SYSC_BASE + (uintptr_t)x1, (uint32_t)x2);
+		SMC_RET0(handle);
+	} else {
+		WARN("%s: Invalid SYSC_PCIe address\n", __func__);
+		SMC_RET1(handle, SMC_ARCH_CALL_INVAL_PARAM);
+	}
+}
+
 uintptr_t rz_plat_sip_handler(uint32_t smc_fid,
 					u_register_t x1,
 					u_register_t x2,
@@ -62,6 +73,8 @@ uintptr_t rz_plat_sip_handler(uint32_t smc_fid,
 		return rz_otp_handler_devid(handle, x1);
 	case RZ_SIP_SVC_GET_CHIPID:
 		return rz_otp_handler_chipid(handle, x1, flags);
+	case RZ_SIP_SVC_SET_PCIE_RST_RSMB:
+		return rz_otp_handler_set_pcie(handle, x1, x2);
 	default:
 		WARN("%s: Unimplemented RZ SiP Service Call: 0x%x\n", __func__, smc_fid);
 		SMC_RET1(handle, SMC_UNK);
diff --git a/plat/renesas/rz/soc/g3s/include/rz_soc_def.h b/plat/renesas/rz/soc/g3s/include/rz_soc_def.h
index 97dd2e7eb..d49e3771b 100755
--- a/plat/renesas/rz/soc/g3s/include/rz_soc_def.h
+++ b/plat/renesas/rz/soc/g3s/include/rz_soc_def.h
@@ -123,6 +123,7 @@
 
 #define RZ_SOC_SYSC_BASE_DEVID			SYS_LSI_DEVID
 #define RZ_SOC_OTP_BASE_CHIPID			(RZG3S_OTP_BASE + 0x1140)
+#define RZG3S_SYSC_PCIE_RST_RSM_B_OFFSET		UL(0x0D74)
 
 #define RZ_SOC_I2C_BASE					RZG3S_I2C_1_BASE
 
-- 
2.25.1

